### PR TITLE
Add maximum quantities for skus

### PR DESCRIPTION
**Pull request by mauricevogel**
This will add the possibility to also specify maximum quantities for SKUs

### CODE REVIEW

File: lib/order_optimizer/catalog.rb [hunk 1]
```diff
@@ -3,7 +3,10 @@
 class OrderOptimizer
   class Catalog
     def initialize(skus)
-      @skus = skus.map { |id, values| Sku.new(id, **values) }.sort_by(&:price_per_unit)
+      @skus = skus
+              .map { |id, values| Sku.new(id, **values) }
+              .reject { |sku| sku.min_quantity > sku.max_quantity if sku.min_quantity && sku.max_quantity }
```
Comment (nickcampbell18) [inline lib/order_optimizer/catalog.rb]: Can we make this explode? User error...

```ruby
class OrderOptimizer::Sku
  def initialize(...)
    @min_quantity = ...
    @max_quantity = ...
    raise ArgumentError, "min_quantity > max_quantity" if min_quantity > max_quantity
```

File: lib/order_optimizer/sku.rb [hunk 2]
```diff
@@ -1,10 +1,11 @@
 class OrderOptimizer
   class Sku
-    attr_reader :id, :quantity, :price_per_unit, :price_per_sku, :min_quantity
+    attr_reader :id, :quantity, :price_per_unit, :price_per_sku, :max_quantity, :min_quantity
 
-    def initialize(id, quantity:, price_per_sku: nil, price_per_unit: nil, min_quantity: nil)
+    def initialize(id, quantity:, price_per_sku: nil, price_per_unit: nil, max_quantity: nil, min_quantity: nil)
       @id = id
       @quantity = BigDecimal(quantity, 2)
+      @max_quantity = BigDecimal(max_quantity, 2) if max_quantity
```
Comment (nickcampbell18) [inline lib/order_optimizer/sku.rb]: Swap this and the line below, `@min` should always come before `@max`.

File: lib/order_optimizer/sku.rb [hunk 3]
```diff
@@ -1,10 +1,11 @@
 class OrderOptimizer
   class Sku
-    attr_reader :id, :quantity, :price_per_unit, :price_per_sku, :min_quantity
+    attr_reader :id, :quantity, :price_per_unit, :price_per_sku, :max_quantity, :min_quantity
 
-    def initialize(id, quantity:, price_per_sku: nil, price_per_unit: nil, min_quantity: nil)
+    def initialize(id, quantity:, price_per_sku: nil, price_per_unit: nil, max_quantity: nil, min_quantity: nil)
```
Comment (nickcampbell18) [inline lib/order_optimizer/sku.rb]: Same here, `min_quantity: nil, max_quantity: nil`

File: lib/order_optimizer.rb [hunk 4]
```diff
@@ -55,12 +55,18 @@ def find_possible_orders(required_qty:, skus:)
   def count_and_remainder_for_sku(quantity, sku)
     count, remainder = quantity.divmod(sku.quantity)
 
-    if sku.min_quantity && count * sku.quantity < sku.min_quantity
+    if sku.max_quantity && count * sku.quantity > sku.max_quantity
```
Comment (nickcampbell18) [inline lib/order_optimizer.rb:58]: I wonder if this lends itself to a range, something like `if sku.quantities.cover?(count * sku.quantity)` where `sku.quantities` is defined as `(min_quantity..max_quantity)`? Could be a nice refactor, as an idea.

**Review by nickcampbell18 — APPROVED** [review]

File: lib/order_optimizer.rb (hunk 4 above)
Comment (mauricevogel) [reply lib/order_optimizer.rb:58]: Not sure. We're not only interested in whether or not we're in the range, but also whether or not we're above or below it.

**Pull request merged**